[RunAllTests] Fix #2900: Add network interceptor to request headers

jcqli opened the pull request:
## Explanation
Fix part of #53
Fix #2900
Implements a new network interceptor that adds request headers to outgoing requests as part of feedback reporting project, although will be applied to all outgoing messages. Renames original `NetworkInterceptor` => `JsonPrefixNetworkInterceptor` to create separate interceptors with independent roles.

[Design doc](https://docs.google.com/document/d/1_RTzad6u4m-ZFn0VTWNddp4bc0fCYrMk-ABIVYVJ7gU/edit?usp=sharing)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

jcqli committed:
Create protos for feedback report

jcqli committed:
Add bool to app startup state for crash

jcqli committed:
Define data objects and service for sending feebdack reports

jcqli committed:
Fix typo

jcqli committed:
Include unspecified for enum types

jcqli committed:
Add protos to model bazel

jcqli committed:
Fix typos

jcqli committed:
Fix typos

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Create data class objets for feedback reporting

jcqli committed:
Create data class objets for feedback reporting

jcqli committed:
Remove topic database from feedback reporting proto

jcqli committed:
Remove topic database from feedback reporting proto

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Finish adding all data class objects

jcqli committed:
Finish adding all data class objects

jcqli committed:
Fix typo

jcqli committed:
Fix typo

jcqli committed:
Clarify comment

jcqli committed:
Clarify comment

jcqli committed:
Apply review comments

jcqli committed:
Apply review comments

jcqli committed:
Clarify proto enums

jcqli committed:
Clarify proto enums

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Update data class models after, add API key to NetworkModule

jcqli committed:
Update data class models after, add API key to NetworkModule

jcqli committed:
Create service mock and fake JSON data for testing

jcqli committed:
Create service mock and fake JSON data for testing

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Write service tests

jcqli committed:
Write service tests

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-data-layer-componets-feedback-reporting

jcqli committed:
Update timestamp data type to be double

jcqli committed:
Update timestamp data type to be double

jcqli committed:
Linter

jcqli committed:
Linter

jcqli committed:
Undo changes to cache store test

jcqli committed:
Undo changes to cache store test

jcqli committed:
Fix linter errors

jcqli committed:
Fix linter errors

jcqli committed:
Remove comma at end of data class Json items

jcqli committed:
Adjust service defintion

jcqli committed:
Merge branch 'create-data-layer-componets-feedback-reporting' of github.com:jcqli/oppia-android into create-data-layer-componets-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'jcqli/create-data-layer-componets-feedback-reporting' into create-data-layer-for-feedback-reporting

jcqli committed:
Create oneof for each issue category

jcqli committed:
Address review comments

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Adjust test to execute and check for success

jcqli committed:
Update data classes to match protobufs

jcqli committed:
Remove unused imports

jcqli committed:
Clarify field names / data

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-feedback-reporting-data-objects

jcqli committed:
Address review comments

jcqli committed:
Clarify locale

jcqli committed:
Apply review comments

jcqli committed:
Rename to AppContext

jcqli committed:
Add typo to issue

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Update GAE models to match protos, fix names, add logs to app context

jcqli committed:
Fix linter error

jcqli committed:
Clarify country locale

jcqli committed:
Update service test to use JSON to make GAE object

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Stub controller

jcqli committed:
Remove controller from this PR

jcqli committed:
Start management controller for feedback reporting

jcqli committed:
Apply review comments

jcqli committed:
Add api key and authorization class

jcqli committed:
Remove api key from network module

jcqli committed:
Add KDoc comments to explain GAE model fields

jcqli committed:
Fix typo in comment

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Fix typo

jcqli committed:
Update URL path for service

jcqli committed:
Add functions to report management controller that fetches / saves reports in local store

jcqli committed:
Provide implementation for feedback reporting service

jcqli committed:
Apply review comments

jcqli committed:
Merge branch 'create-data-layer-for-feedback-reporting' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix typo

jcqli committed:
Add service call to controller

jcqli committed:
Remove api key from moshi object to use in interceptor for request header

jcqli committed:
Adjust comment

jcqli committed:
Change JSON field name to match backend

jcqli committed:
Implement cases for different issues for  user supplied info

jcqli committed:
Add context info to controller

jcqli committed:
Get event logs

jcqli committed:
Fix error message

jcqli committed:
Add headers using network interceptor

jcqli committed:
Create event logs

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Resolve conflicts

jcqli committed:
Change topic_progress to completed_exploration_ids

jcqli committed:
Fix comment

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Add entry point as a data class

jcqli committed:
Add entry point in controller

jcqli committed:
Start tests

jcqli committed:
Rename original interceptor into json prefix and change references

jcqli committed:
Separate header function for testing

jcqli committed:
Separate header function for testing

jcqli committed:
Merge branch 'add-network-interceptor-to-request-headers' of github.com:oppia/oppia-android into add-network-interceptor-to-request-headers

jcqli committed:
Inject context in interceptor

jcqli committed:
Add test cases for adding and replacing headers

jcqli committed:
Merge branch 'develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Merge branch 'create-data-layer-for-feedback-reporting' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix linter errors

jcqli committed:
Add comments to clarify controller

jcqli committed:
Configure bazel?'

jcqli committed:
Merge branch 'create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Add annotation class for network api key

jcqli committed:
Bazel build?

anandwana001 commented:
@jcqli Setting the script will help you prevent kotlin lint failures. - https://github.com/oppia/oppia-android/wiki#installation

jcqli committed:
Clarify country code for locales

jcqli committed:
Specify language code'

jcqli committed:
Fix names to match JSON

jcqli committed:
Mock retrofit and service in test

jcqli committed:
Move mock services to testing module

jcqli committed:
Update data fields to match backend

jcqli committed:
Read logcat logs from file

jcqli committed:
Test builds

jcqli committed:
Remove completed explorations list

jcqli committed:
Make fake file for logcat

jcqli committed:
Init mocks in test

jcqli committed:
Update codeowners and fix linter errors

jcqli committed:
Undo bazel changes

jcqli committed:
Move mock services and api utils to testing module

jcqli committed:
Merge remote-tracking branch 'origin/create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Remove extra newline

jcqli committed:
Change field names to match backend

jcqli committed:
Use class const for language codes

jcqli committed:
Explain logcat file helper in a comment

jcqli committed:
Fix tests

jcqli committed:
Merge remote-tracking branch 'origin/create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Fix bazel build

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Add moshi for annotations in bazel?

jcqli committed:
Fix injects in test

jcqli committed:
Remove moshi from bazel

jcqli committed:
Change audio language code to audio language

jcqli committed:
Update data classes to match backend declarations

jcqli committed:
Fix linter errors

jcqli committed:
Use Callback for request

jcqli committed:
Update field to match JSON

jcqli committed:
Fix linter errors

jcqli committed:
Fix linter errors

jcqli committed:
Create language code extensions

jcqli committed:
Linter fixes

jcqli committed:
Merge remote-tracking branch 'origin/develop' into add-network-interceptor-to-request-headers

jcqli committed:
Merge remote-tracking branch 'origin/create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Use Call in service, change timestamp to int for seconds

jcqli committed:
Fix JSON test errors

jcqli committed:
Cleanup comments, code

jcqli committed:
Fix codeowners for JSON in data module

jcqli committed:
Move JSON to data module

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'origin/develop' into add-network-interceptor-to-request-headers

jcqli committed:
Merge branch 'develop' into add-network-interceptor-to-request-headers

BenHenning committed:
Enable Moshi support in Bazel.

This includes some reorganizing of the data module, but it introduces
support for generating GAE models through Moshi. Corresponding mock
services & tests are still failing, but //data & //:oppia build.

BenHenning committed:
Make mock services build for testing module.

BenHenning committed:
Add working data tests to Bazel build.

This also fixes a regression from an earlier commit that led to the
Moshi objects not actually being generated correctly, so the data
service tests are good proof that Moshi is being used correctly.

BenHenning committed:
Remove TODOs for #1675.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into enable-data-module-in-bazel

Conflicts:
	.bazelrc

BenHenning committed:
Bazel lint fix.

Avoid relying on native java_library.

BenHenning committed:
Fix broken Gradle build.

This includes explicitly specifying okhttp version since Gradle was
previously randomly picking one.

BenHenning committed:
Add debug details for affected tests script.

BenHenning committed:
Lint fixes.

BenHenning committed:
Attempt to force Bazel cache to clear.

This is an attempt to work around a theorized OOM happening when trying
to compute the sibling build graph for changes build files.

jcqli committed:
Merge branch 'develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge branch 'enable-data-module-in-bazel' into create-domain-controller-for-feedback-reporting

jcqli committed:
Create module for report schema version

jcqli committed:
Add tests for schema validator

jcqli committed:
Add FeedbackReportingModule to the application component

jcqli committed:
Adjust injection for schema version, tests build and run

jcqli committed:
Add test to check fields of high-level feedback report

jcqli committed:
Add field checks for all data classes

jcqli committed:
Use helper to verify data classes

jcqli committed:
Add entry point to model bazel, remove items from domain bazel

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Add bazel dependencies

jcqli committed:
Make when statement exhaustive with no-op blocks

jcqli committed:
Remove imports and udpate JSON field names

jcqli committed:
Nit fixes

jcqli committed:
Merge branch 'create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Fix nits

jcqli committed:
Merge branch 'develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge branch 'develop' into add-network-interceptor-to-request-headers

jcqli committed:
Merge branch 'create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Use JVM annotation to inject api key

jcqli committed:
Nit fixes

jcqli committed:
Merge branch 'create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Update POST url

jcqli committed:
Add newly-named interceptors to bazel build

jcqli committed:
Use ContextExtensions to get package info

jcqli committed:
Set up test application in test

jcqli committed:
Use application component provider

jcqli commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,170 @@
+        .setPackageName("test_package_name")
+        .setApplicationInfo(applicationInfo)
+        .build()
+    packageInfo.versionName = "1.0"
+    packageInfo.versionCode = 1
+    packageManager.installPackage(packageInfo)
```
This doesn't seem to set the versionName or versionCode -- it throws a NullPointerException later on in the test for the versionName when the RemoteAuthNetworkInterceptor tries to set the headers ([line 44)](https://github.com/oppia/oppia-android/pull/2907/files#diff-c443970e72f2a145c0321816f33ec349b2cdf2054e8bc3b2f0338cb742ef2cf2R44). Any tips?

BenHenning replied:
I think NPE can happen in Context.getVersionName() if there's no PackageInfo for the current package name. You should be able to debug in this test what the package name is for Context and whether it's matching the PackageInfo you're creating (which is using package name 'test_package_name').

jcqli committed:
Fix typo

jcqli committed:
Add extensions to view model bazel

jcqli committed:
Set test application

jcqli committed:
Use context values for testing header

jcqli committed:
Fix linter issues

jcqli committed:
Clean up test

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Remove duplicate dependency naming

jcqli committed:
Add MockWebServer to controller test

jcqli committed:
Fix linter errors

BenHenning commented on utility/src/main/java/org/oppia/android/util/extensions/BUILD.bazel:
```diff
@@ -16,7 +16,7 @@ kt_android_library(
 
 kt_android_library(
-    name = "context_extensinos",
+    name = "context_extensions",
```
Nice catch.

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt:
```diff
@@ -17,7 +17,8 @@ import javax.inject.Singleton
  */
 @Module
-class NetworkModule {
+class
+NetworkModule {
```
Should be on the same line.

jcqli replied:
Done, I saw this as well but it didn't show up on my AS until now. How odd

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt:
```diff
@@ -31,7 +32,8 @@ class NetworkModule {
   fun provideRetrofitInstance(): Retrofit {
     val client = OkHttpClient.Builder()
-    client.addInterceptor(NetworkInterceptor())
+    client.addInterceptor(JsonPrefixNetworkInterceptor())
+      .addInterceptor(RemoteAuthNetworkInterceptor())
```
These should be parameter injected, not constructed, otherwise their injections won't be ready.

You can do that by adding these as parameters to the provider method.

jcqli replied:
Done

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt:
```diff
@@ -68,4 +70,11 @@ class NetworkModule {
+  // Provides the API key to use in authenticating remote messages sent or received. This will be
+  // replaced with a stronger, secret key will in production.
+  @Provides
+  @Singleton
```
You should remove this--immutable strings don't need to be scoped since they can be trivially re-retrieved.

jcqli replied:
Done

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt:
```diff
@@ -68,4 +70,11 @@ class NetworkModule {
   }
+
+  // Provides the API key to use in authenticating remote messages sent or received. This will be
+  // replaced with a stronger, secret key will in production.
```
```suggestion
  // replaced with a secret key in production.
```

jcqli replied:
Done

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt:
```diff
@@ -0,0 +1,48 @@
+ *
+ * The Interceptor adds header parameters to outgoing messages.
+ */
+@Singleton
```
This class is stateless so it shouldn't be scoped.

jcqli replied:
Done

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt:
```diff
@@ -0,0 +1,48 @@
+   * @param request the request being sent over the network
+   * @return the exact same request as the parameter with added headers
+   */
+  fun addAuthHeaders(request: Request): Request {
```
Can this be private? It seems to only be needed for the class implementation.

jcqli replied:
Is there a way to declare this as test-accessible? The network interceptor tests use this to verify the correct headers are added.

jcqli replied:
I made this an internal fun so that the associated test could call on it and verify that headers are added correctly. I think another way to test this and keep the function private would be calling `intercept()` on a chain and verifying the response's associated request matched the interceptor's changes, but this is a bit more indirect (and probably more difficult) and I'm not sure if it's an accurate representation of the request anymore after it goes through the request. Using an internal function seems to be the most beneficial, WDYT?

BenHenning replied:
It looks like this is resolved now.

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt:
```diff
@@ -27,7 +27,7 @@ class ClassroomServiceTest {
   fun setUp() {
     val client = OkHttpClient.Builder()
-    client.addInterceptor(NetworkInterceptor())
+    client.addInterceptor(JsonPrefixNetworkInterceptor())
```
Please also inject this everywhere it's used in tests, too.

Though, it might be helpful to wait for #2642 if we can since I think that's largely simplifying data test setup.

jcqli replied:
Is this interceptor needed for the mock retrofit? All the service tests pass when removing it, maybe that's an indicator that our mock retrofit is not set up properly.

I ask since injecting the interceptors means we then need to declare TestApplications for all the tests that use the interceptors, is that right?

jcqli replied:
Based on our discussion offline I'm going to leave the interceptor here so that it doesn't interfere with #2642 since it seems that the interceptors will be added there.

BenHenning replied:
Moved this comment to #2916 via https://github.com/oppia/oppia-android/pull/2916/files#r610890818.

BenHenning commented on utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt:
```diff
@@ -9,6 +9,11 @@ fun Context.getVersionName(): String {
 
+/** Returns the current app version code. */
+fun Context.getVersionCode(): Int {
+  return this.packageManager.getPackageInfo(this.packageName, /* flags=*/ 0).versionCode
```
```suggestion
  return this.packageManager.getPackageInfo(this.packageName, /* flags= */ 0).versionCode
```

Maybe fix the one above, too, for an opportunistic fix.

jcqli replied:
Done

BenHenning reviewed (changes requested):
Thanks @jcqli! Left some thoughts. Not sure if you actually want to block on #2916 (might be better not to give the time sensitivity here), but wanted to mention it as an option.

jcqli committed:
Specify language code to string, add profile proto to utility bazel

jcqli committed:
Apply review comments

jcqli committed:
Fix import

jcqli committed:
Merge branch 'create-domain-controller-for-feedback-reporting' into add-network-interceptor-to-request-headers

jcqli committed:
Make function internal for testing use

jcqli committed:
Fix build linter errors

jcqli commented:
> Thanks @jcqli! Left some thoughts. Not sure if you actually want to block on #2916 (might be better not to give the time sensitivity here), but wanted to mention it as an option.

I think I'd rather have this merged and not blocked on #2916 but would need to make some changes so that its' not blocked by #2878, let me know if that would be better (I think it would make more sense, but requires some more work on my end).

Unrelated, I'm not sure why the linter is giving errors on the third_party/versions.bzl file (from 2878 changes). I think it's about how I added the MockWebServer to dependency but this can be resolved in the other PR, and also depends on if we block on 2878. Let me know your thoughts.

jcqli committed:
Remove network interceptors from services tests

jcqli committed:
Revert "Remove network interceptors from services tests"

This reverts commit 18c958f5387f485de5873d5893aec94253eec9fe.

jcqli committed:
Remove network interceptor from feedback reporting service test

jcqli committed:
Hook up mock web server to test

jcqli committed:
Remove domain controller changes from this branch

jcqli committed:
Update data objects to remove controller changes

jcqli committed:
Add comma

jcqli committed:
Update data objects

jcqli committed:
Name interceptor in json prefix test

jcqli committed:
Merge remote-tracking branch 'origin/develop' into add-network-interceptor-to-request-headers

jcqli committed:
Start server

jcqli committed:
Use guava executors?

jcqli committed:
Call okhttp directly for service

jcqli committed:
Use okhttp instead of retrofit for tests

jcqli committed:
Add retrofit to second test case

jcqli committed:
Use retrofit again

prayutsu commented on third_party/versions.bzl:
```diff
@@ -85,6 +85,7 @@ MAVEN_TEST_DEPENDENCY_VERSIONS = {
     "com.github.bumptech.glide:mocks": "4.11.0",
     "com.google.truth:truth": "0.43",
+    "com.squareup.okhttp3:mockwebserver":"4.1.0",
     "com.squareup.retrofit2:retrofit-mock": "2.5.0",
```
I think lint is failing because there's a space missing before the version (after the colon).

BenHenning replied:
Nice catch.

jcqli committed:
Fix missing space

jcqli committed:
Don't use mock retrofit

jcqli committed:
Use mock web server instead of mock retrofit for both interceptor tests

jcqli committed:
Make function private

jcqli commented:
To properly test the network interceptor without exposing private functions (e.g. actually testing the full network request was successful), we used MockWebServer to verify the request call. Turns out, MockWebServer is incompatible with MockRetrofit so the test here does not use the mock services but rather the full service on the mocked server.

/cc @prayutsu Not sure if this will affect your data module PR but might be relevant

jcqli committed:
Fix data class names in test

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest.kt:
```diff
@@ -109,7 +109,7 @@ class JsonPrefixNetworkInterceptorTest {
-    fun inject(networkInterceptorTest: JsonPrefixNetworkInterceptorTest)
-    fun inject(networkInterceptor: JsonPrefixNetworkInterceptor)
+    fun inject(jsonPrefixNetworkInterceptorTest: JsonPrefixNetworkInterceptorTest)
+    fun inject(jsonPrefixNetworkInterceptor: JsonPrefixNetworkInterceptor)
```
This should be removed--only the test needs an injector generated.

jcqli replied:
Done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+    setUpApplicationForContext()
+    setUpRetrofit()
```
I suggest moving your test helpers to also be defined in the same order they're called here (& after the tests since we generally follow an "appendix" style organization for helpers rather than "footnote" style).

jcqli replied:
Ack, done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+    assertThat(serviceRequest.header("app_version_code")).isNull()
+
+    call.execute()
+    val interceptedRequest = mockWebServer.takeRequest(timeout = 8, unit = TimeUnit.SECONDS)
```
Suggest going with 10 seconds for parity with TestCoroutineDispatcher (in fact you could use the fields from there for the time & unit if you'd prefer).

jcqli replied:
Done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+
+    call.execute()
+    val interceptedRequest = mockWebServer.takeRequest(timeout = 8, unit = TimeUnit.SECONDS)
+    verifyRequestHeaders(interceptedRequest!!.headers)
```
```suggestion
    verifyRequestHeaders(interceptedRequest?.headers)
```

Maybe use nullable calls, instead. The assertion should still fail if the individual retrievals are null, though also recommending a non-null assertion for the request headers in verifyRequestHeaders to make failures a bit more explicit.

'!!' isn't quite as ideal since it will throw a NPE rather than AssertionError. It'll still be clear once someone digs into the failure, but test frameworks generally expect AssertionErrors and may provide slightly better error messaging in those cases. Plus, it's just nice to never use '!!' :)

jcqli replied:
Ack, done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+  }
+
+  private fun setUpRetrofit() {
+    mockWebServer.dispatcher
```
This isn't doing anything--I suggest removing it.

jcqli replied:
Ah oops thanks! Done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+
+  private val topicName = "Topic1"
+
+  private val mockWebServer = MockWebServer()
```
Should this be initialized in setUpRetrofit, instead?

Also, does it need to be torn down in an ``@After`` function?

jcqli replied:
Yup, done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,207 @@
+
+  private val topicName = "Topic1"
+
+  private val mockWebServer = MockWebServer()
```
Should this be initialized in setUpRetrofit, instead?

Also, does it need to be torn down in an ``@After`` function?

jcqli replied:
Done

jcqli committed:
Remove test provider for mock service

BenHenning reviewed (changes requested):
Thanks @jcqli. Not sure why my previous review was submitted automatically, but either way--just have a few comments to address but otherwise I think the PR looks good.

BenHenning commented:
@jcqli you may also want to assign the other code owners so that they can take a look.

jcqli committed:
Shut down mock server

jcqli commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,200 @@
+    val call = topicService.getTopicByName(topicName)
+
+    val serviceRequest = topicService.getTopicByName(topicName).request()
+    val request = serviceRequest.newBuilder()
+      .addHeader("api_key", "wrong_api_key")
+      .addHeader("app_package_name", "wrong_package_name")
+      .addHeader("app_version_name", "wrong_version_name")
+      .addHeader("app_version_code", "wrong_version_code")
+      .build()
```
@BenHenning I realized that this test doesn't actually use the request with the incorrectly set headers. I don't believe MockWebServer enables us to enqueue a specific request, nor do I know how to pass a request directly into the service (unless we want to mock the service differently). Do you have any suggestions on how we could test this case?

BenHenning replied:
Per offline discussion I think we landed on using Okhttp directly (& per the latest code)--can this be resolved now?

jcqli replied:
Ah yup, thanks for following-up on this

jcqli committed:
Add not null assertion, set up server with retrofit

jcqli committed:
Call execute on okhttp client to specify request

jcqli commented:
> @jcqli you may also want to assign the other code owners so that they can take a look.

Ack, done!

rt4914 commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkApiKey.kt:
```diff
@@ -0,0 +1,7 @@
+
+import javax.inject.Qualifier
+
+/** Qualifier for the API key to use when validating incoming and outgoing requests */
```
Nit: `.` at the end.

jcqli replied:
Done

rt4914 commented on data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt:
```diff
@@ -29,9 +29,13 @@ class NetworkModule {
+  ): Retrofit {
     val client = OkHttpClient.Builder()
-    client.addInterceptor(NetworkInterceptor())
+    client.addInterceptor(jsonPrefixNetworkInterceptor)
```
Suggest changing this code to either one of these
```
    val clientBuilder = OkHttpClient.Builder()
      .addInterceptor(jsonPrefixNetworkInterceptor)
      .addInterceptor(remoteAuthNetworkInterceptor)
```

or 

```
    val client = OkHttpClient.Builder()
      .addInterceptor(jsonPrefixNetworkInterceptor)
      .addInterceptor(remoteAuthNetworkInterceptor)
      .build()
```

Because currently your object `client` is actually a `clientBuilder`

jcqli replied:
Done

rt4914 commented on data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt:
```diff
@@ -29,7 +28,6 @@ class FeedbackReportingServiceTest {
   @Before
   fun setUp() {
     val client = OkHttpClient.Builder()
-    client.addInterceptor(NetworkInterceptor())
```
Does `FeedbackReportingService` do not need `JsonPrefixNetworkInterceptor ` ?

jcqli replied:
The service currently doesn't receive anything in the response, so I don't believe so since the interceptor only handles responses.

rt4914 replied:
Ack. Thanks.

rt4914 reviewed (approved):
LGTM, there are three comments. Two you can resolve and other one you can answer and I will check it. Thanks.

anandwana001 reviewed (approved):
test files LGTM

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -92,51 +89,82 @@ class RemoteAuthNetworkInterceptorTest {
+      timeout = testCoroutineDispatcher.DEFAULT_TIMEOUT_SECONDS,
+      unit = testCoroutineDispatcher.DEFAULT_TIMEOUT_UNIT
+    )
+    interceptedRequest?.let {
+      verifyRequestHeaders(it.headers)
+    }
```
I was thinking more like: verifyRequestedHeaders(it?.headers). The null check should work automatically inside verify since you'll also need to do the safe-call operator & Truth will fail if the resulting expression evaluates to null.

jcqli replied:
Done

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -92,51 +89,82 @@ class RemoteAuthNetworkInterceptorTest {
-    val serviceRequest = topicService.getTopicByName(topicName).request()
-    val request = serviceRequest.newBuilder()
+  fun testNetworkInterceptor_withIncorrectHeaders_setsCorrectHeaders() {
+    val request = Request.Builder()
+      .url(mockWebServer.url("/"))
```
Maybe explain briefly via comment why we can't use Retrofit in this test case.

jcqli replied:
Done

BenHenning reviewed (commented):
Thanks @jcqli! Generally LGTM, but can you re-run the Bazel tests? I think you're running into one of the known CI issues where compute_affected_tests hangs. You'll need to add '[RunAllTests]' to the PR title & then re-run the Bazel tests workflow.

jcqli committed:
Apply review comments

jcqli committed:
Merge branch 'develop' into add-network-interceptor-to-request-headers

jcqli committed:
Apply review comments

jcqli committed:
Add null-type check for headers

jcqli committed:
Fix dependency

BenHenning reviewed (commented):
Thanks @jcqli! Looks like you need to update TestDispatcherModule & RobolectricModule imports in your new tests--the builds for those tests are failing on both Bazel & Gradle. In general, it's good practice to re-run key tests locally (such as new/changed suites) after a merge just to double-check integrations post-sync.

PR LGTM otherwise.

jcqli committed:
Merge remote-tracking branch 'origin/develop' into add-network-interceptor-to-request-headers

jcqli committed:
Update module paths

jcqli commented:
> Thanks @jcqli! Looks like you need to update TestDispatcherModule & RobolectricModule imports in your new tests--the builds for those tests are failing on both Bazel & Gradle. In general, it's good practice to re-run key tests locally (such as new/changed suites) after a merge just to double-check integrations post-sync.

@BenHenning Thanks for taking a look!

jcqli committed:
Add mockwebserver to data bazel build

BenHenning reviewed (changes requested):
Thanks @jcqli. ``RemoteAuthNetworkInterceptorTest`` appears broken on Bazel. Please fix & send back. (I highly encourage you verify that test builds & passes locally before sending back since the CI run will take ~2 hours to run).

jcqli committed:
Fix build for data tests

jcqli commented:
> Thanks @jcqli. `RemoteAuthNetworkInterceptorTest` appears broken on Bazel. Please fix & send back. (I highly encourage you verify that test builds & passes locally before sending back since the CI run will take ~2 hours to run).

Aw man, I missed the data module -- thanks for the pointer, it should be good now!

BenHenning reviewed (approved):
Looks like everything is passing now--LGTM. Thanks @jcqli!
